File: packages/core/tools/matimo_get_skill/matimo_get_skill.ts
import fs from 'fs';
import path from 'path';
import { getGlobalMatimoLogger, getGlobalMatimoInstance, ToolLoader } from '@matimo/core';
import { parseSkillContent, listBundledResources, type BundledResources } from '../shared/skill-validation';

interface GetSkillParams {
  name: string;
  skills_dir?: string;
  file?: string;
}

interface GetSkillResult {
  success: boolean;
  name?: string;
  description?: string;
  content?: string;
  path?: string;
  message: string;
  license?: string;
  compatibility?: string;
  metadata?: Record<string, string>;
  resources?: BundledResources;
}

/** Path traversal detection — defense-in-depth. */
const UNSAFE_NAME_PATTERN = /[/\\]|\.\.|[\x00-\x1f]/;

/**
 * Helper: Find skill directory using auto-discovery (like matimo_list_skills)
 */
function findSkillDir(skillName: string, explicitSkillsDir?: string): string | null {
  // Try explicit skills_dir first
  if (explicitSkillsDir) {
    const skillPath = path.join(explicitSkillsDir, skillName, 'SKILL.md');
    if (fs.existsSync(skillPath)) {
      return path.join(explicitSkillsDir, skillName);
    }
  }

  // Try MatimoInstance
  try {
    const matimo = getGlobalMatimoInstance();
    if (matimo) {
      const skills = matimo.listSkills();
      const found = skills?.find((s) => s.name === skillName);
      if (found) {
        const skillPath = (found as any)._path;
        if (skillPath && fs.existsSync(path.join(skillPath, 'SKILL.md'))) {
          return skillPath;
        }
      }
    }
  } catch {
    // Fall through
  }

  // Auto-discover from @matimo/* packages
  try {
    const toolLoader = new ToolLoader();
    const discoveredPaths = toolLoader.autoDiscoverPackages();
    for (const toolPath of discoveredPaths) {
      const pkgDir = path.dirname(toolPath);
      const skillPath = path.join(pkgDir, 'skills', skillName, 'SKILL.md');
      if (fs.existsSync(skillPath)) {
        return path.join(pkgDir, 'skills', skillName);
      }
    }
  } catch {
    // Fall through
  }

  return null;
}

/**
 * Read a skill's content by name — Level 2 activation (SKILL.md) or
 * Level 3 resource access (bundled file).
 *
 * When called without `file`, returns SKILL.md content + metadata + resource listing.
 * When called with `file`, returns the contents of that bundled resource file.
 *
 * Skills are discovered in this order (priority):
 * 1. Explicit skills_dir if provided
 * 2. Global MatimoInstance (if initialized)
 * 3. Auto-discovered @matimo/* packages
 *
 * @see https://agentskills.io/specification
 */
export default async function matimoGetSkill(
  params: GetSkillParams,
): Promise<GetSkillResult> {
  const logger = getGlobalMatimoLogger();

  if (!params.name || params.name.trim().length === 0) {
    return { success: false, message: 'Skill name is required' };
  }

  if (UNSAFE_NAME_PATTERN.test(params.name)) {
    return { success: false, message: 'Skill name contains invalid characters' };
  }

  // Find skill directory using auto-discovery
  const skillDir = findSkillDir(params.name, params.skills_dir);
  if (!skillDir) {
    return { success: false, message: `Skill "${params.name}" not found` };
  }

  const skillPath = path.join(skillDir, 'SKILL.md');

  // Level 3: Read a specific bundled resource file
  if (params.file) {
    // For file paths, allow forward slashes but reject path traversal
    if (/\.\.|\\/u.test(params.file) || /[\x00-\x1f]/.test(params.file)) {
      return { success: false, message: 'File path contains invalid characters' };
    }

    const resourcePath = path.join(skillDir, params.file);
    // Verify the resolved path stays within the skill directory
    const resolvedPath = path.resolve(resourcePath);
    const resolvedSkillDir = path.resolve(skillDir);
    if (!resolvedPath.startsWith(resolvedSkillDir + path.sep) && resolvedPath !== resolvedSkillDir) {
      return { success: false, message: 'File path escapes the skill directory' };
    }

    if (!fs.existsSync(resourcePath)) {
      return { success: false, message: `Resource file "${params.file}" not found in skill "${params.name}"` };
    }

    try {
      const fileContent = fs.readFileSync(resourcePath, 'utf-8');
      return {
        success: true,
        name: params.name,
        content: fileContent,
        path: resourcePath,
        message: `Resource file "${params.file}" retrieved successfully.`,
      };
    } catch (err) {
      return { success: false, message: `Failed to read resource file: ${(err as Error).message}` };
    }
  }

  // Level 2: Read SKILL.md + metadata + resource listing
  try {
    const rawContent = fs.readFileSync(skillPath, 'utf-8');
    const parseResult = parseSkillContent(rawContent);

    const result: GetSkillResult = {
      success: true,
      name: params.name,
      content: rawContent,
      path: skillPath,
      message: 'Skill retrieved successfully.',
    };

    if (parseResult.success && parseResult.parsed) {
      const { frontmatter } = parseResult.parsed;
      result.name = frontmatter.name || params.name;
      result.description = frontmatter.description || '';
      if (frontmatter.license) result.license = frontmatter.license;
      if (frontmatter.compatibility) result.compatibility = frontmatter.compatibility;
      if (frontmatter.metadata) result.metadata = frontmatter.metadata;
    }

    // List bundled resources (Level 3 discovery)
    result.resources = listBundledResources(skillDir);

    logger.info('matimo_get_skill: skill retrieved', {
      name: params.name,
      path: skillPath,
    });

    return result;
  } catch (err) {
    const errorMsg = (err as Error).message;
    logger.error('matimo_get_skill: failed to read skill', {
      name: params.name,
      error: errorMsg,
    });
    return { success: false, message: `Failed to read skill: ${errorMsg}` };
  }
}

File: python/packages/core/src/matimo/tools/matimo_get_skill/matimo_get_skill.py
"""matimo_get_skill — read a skill's SKILL.md content and metadata."""
from __future__ import annotations

import logging
import re
from pathlib import Path

import yaml

logger = logging.getLogger("matimo")

UNSAFE_NAME = re.compile(r"[/\\]|\.\.|[\x00-\x1f]")


def _extract_frontmatter(content: str) -> tuple[dict, str]:  # type: ignore[type-arg]
    if not content.startswith("---"):
        return {}, content
    end = content.find("---", 3)
    if end == -1:
        return {}, content
    try:
        fm: dict = yaml.safe_load(content[3:end]) or {}  # type: ignore[type-arg]
    except Exception:
        fm = {}
    body = content[end + 3:].lstrip("\n")
    return fm, body


def _scan_resources(skill_dir: Path) -> dict:  # type: ignore[type-arg]
    resources: dict = {"scripts": [], "references": [], "assets": [], "other": []}  # type: ignore[type-arg]
    for entry in sorted(skill_dir.iterdir()):
        if entry.name == "SKILL.md":
            continue
        if entry.is_file():
            resources["other"].append(entry.name)
        elif entry.is_dir():
            if entry.name in ("scripts", "references", "assets"):
                resources[entry.name] = [f.name for f in sorted(entry.iterdir()) if f.is_file()]
    return resources


def _find_skill_dir(name: str, explicit_dir: str | None) -> Path | None:
    if explicit_dir:
        p = Path(explicit_dir) / name
        if (p / "SKILL.md").exists():
            return p

    # Try global instance skill paths
    try:
        from matimo.decorators import get_global_matimo_instance

        instance = get_global_matimo_instance()
        if instance is not None and hasattr(instance, "list_skills"):
            for s in instance.list_skills():
                if s.name == name and getattr(s, "_path", None):
                    p = Path(s._path)  # type: ignore[union-attr]
                    if (p / "SKILL.md").exists():
                        return p
    except Exception as exc:
        logger.debug("matimo_get_skill: global instance lookup failed: %s", exc)

    return None


async def run(params: dict) -> dict:  # type: ignore[type-arg]
    name: str = (params.get("name") or "").strip()
    skills_dir: str | None = params.get("skills_dir")
    file: str | None = params.get("file")

    if not name:
        return {"success": False, "message": "Skill name is required"}
    if UNSAFE_NAME.search(name):
        return {"success": False, "message": "Skill name contains invalid characters"}

    skill_dir = _find_skill_dir(name, skills_dir)
    if skill_dir is None:
        return {"success": False, "name": name, "message": f'Skill "{name}" not found'}

    if file:
        # Reject path traversal in the file param
        if re.search(r"\.\.|\\|[\x00-\x1f]", file):
            return {"success": False, "message": "File path contains invalid characters"}
        resource_path = skill_dir / file
        # Ensure the resolved path stays inside the skill directory
        resolved = resource_path.resolve()
        skill_dir_resolved = skill_dir.resolve()
        if resolved != skill_dir_resolved and not resolved.is_relative_to(skill_dir_resolved):
            return {"success": False, "message": "File path escapes the skill directory"}
        if not resource_path.exists():
            return {
                "success": False,
                "name": name,
                "message": f'Resource file "{file}" not found in skill "{name}"',
            }
        return {
            "success": True,
            "name": name,
            "content": resource_path.read_text(encoding="utf-8"),
            "path": str(resource_path),
            "message": f'Resource file "{file}" retrieved successfully.',
        }

    skill_file = skill_dir / "SKILL.md"
    content = skill_file.read_text(encoding="utf-8")
    fm, _body = _extract_frontmatter(content)
    resources = _scan_resources(skill_dir)

    return {
        "success": True,
        "name": fm.get("name", name),
        "description": fm.get("description", ""),
        "content": content,
        "path": str(skill_file),
        "license": fm.get("license"),
        "compatibility": fm.get("compatibility"),
        "metadata": fm.get("metadata"),
        "resources": resources,
        "message": "Skill retrieved successfully.",
    }
